src/topology/ping_pong.rs

```diff
@@ -0,0 +1,865 @@
+    ///
+    ///
+    /// [VDAF]: https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-vdaf-06#section-5.8
+    /// [issue]: https://github.com/divviup/libprio-rs/issues/670
+    fn continued(
+        &self,
+        role: Role,
+        host_state: Self::State,
+        inbound: &Message,
+    ) -> Result<Self::ContinuedValue, PingPongError>;
+}
+
+impl<const VERIFY_KEY_SIZE: usize, const NONCE_SIZE: usize, A>
+    PingPongTopology<VERIFY_KEY_SIZE, NONCE_SIZE> for A
+where
+    A: Aggregator<VERIFY_KEY_SIZE, NONCE_SIZE>,
+{
+    type State = State<VERIFY_KEY_SIZE, NONCE_SIZE, Self>;
+    type ContinuedValue = ContinuedValue<VERIFY_KEY_SIZE, NONCE_SIZE, Self>;
+    type Transition = Transition<VERIFY_KEY_SIZE, NONCE_SIZE, Self>;
+
+    fn leader_initialize(
+        &self,
+        verify_key: &[u8; VERIFY_KEY_SIZE],
+        agg_param: &Self::AggregationParam,
+        nonce: &[u8; NONCE_SIZE],
+        public_share: &Self::PublicShare,
+        input_share: &Self::InputShare,
+    ) -> Result<(Self::State, Message), PingPongError> {
+        self.prepare_init(
+            verify_key,
+            /* Leader */ 0,
+            agg_param,
+            nonce,
+            public_share,
+            input_share,
+        )
+        .map(|(prep_state, prep_share)| {
+            (
+                State::Continued(prep_state),
+                Message::Initialize {
+                    prep_share: prep_share.get_encoded(),
+                },
+            )
+        })
+        .map_err(PingPongError::VdafPrepareInit)
+    }
+
+    fn helper_initialize(
+        &self,
+        verify_key: &[u8; VERIFY_KEY_SIZE],
+        agg_param: &Self::AggregationParam,
+        nonce: &[u8; NONCE_SIZE],
+        public_share: &Self::PublicShare,
+        input_share: &Self::InputShare,
+        inbound: &Message,
+    ) -> Result<Self::Transition, PingPongError> {
+        let (prep_state, prep_share) = self
+            .prepare_init(
+                verify_key,
+                /* Helper */ 1,
+                agg_param,
+                nonce,
+                public_share,
+                input_share,
+            )
+            .map_err(PingPongError::VdafPrepareInit)?;
+
+        let inbound_prep_share = if let Message::Initialize { prep_share } = inbound {
+            Self::PrepareShare::get_decoded_with_param(&prep_state, prep_share)
+                .map_err(PingPongError::CodecPrepShare)?
+        } else {
+            return Err(PingPongError::StateMismatch(
+                "initialize",
+                inbound.state_name(),
+            ));
+        };
+
+        let current_prepare_message = self
+            .prepare_preprocess([inbound_prep_share, prep_share])
+            .map_err(PingPongError::VdafPreparePreprocess)?;
+
+        Ok(Transition {
+            previous_prepare_state: prep_state,
+            current_prepare_message,
+        })
+    }
+
+    fn continued(
+        &self,
+        role: Role,
```

Comment:
It's a bit awkward to use `Role` to distinguish roles here, but for initialization we the name of the method we call is used to distinguish. We should be consistent.

My suggestion would be to split `continued` into `leader_continued` and `helper_continued`, that way we don't have to expose `Role` in the public API.